![New York City schoolbus](schoolbus.jpg)

Photo by [Jannis Lucas](https://unsplash.com/@jannis_lucas) on [Unsplash](https://unsplash.com).
<br>

Every year, American high school students take SATs, which are standardized tests intended to measure literacy, numeracy, and writing skills. There are three sections - reading, math, and writing, each with a **maximum score of 800 points**. These tests are extremely important for students and colleges, as they play a pivotal role in the admissions process.

Analyzing the performance of schools is important for a variety of stakeholders, including policy and education professionals, researchers, government, and even parents considering which school their children should attend. 

You have been provided with a dataset called `schools.csv`, which is previewed below.

You have been tasked with answering three key questions about New York City (NYC) public school SAT performance.

In [88]:
# Re-run this cell 
import pandas as pd

# Read in the data
schools = pd.read_csv("schools.csv")

# Preview the data
schools.head()

# Start coding here...
# Add as many cells as you like...

school_name  ... percent_tested
0  New Explorations into Science, Technology and ...  ...            NaN
1                               Essex Street Academy  ...           78.9
2                       Lower Manhattan Arts Academy  ...           65.1
3    High School for Dual Language and Asian Studies  ...           95.9
4      Henry Street School for International Studies  ...           59.7

[5 rows x 7 columns]

In [89]:
# let us explore our dataset 

# The columns:

schools.columns

Index(['school_name', 'borough', 'building_code', 'average_math',
       'average_reading', 'average_writing', 'percent_tested'],
      dtype='object')

In [90]:
# The information about our columns 

schools.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 375 entries, 0 to 374
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   school_name      375 non-null    object 
 1   borough          375 non-null    object 
 2   building_code    375 non-null    object 
 3   average_math     375 non-null    int64  
 4   average_reading  375 non-null    int64  
 5   average_writing  375 non-null    int64  
 6   percent_tested   355 non-null    float64
dtypes: float64(1), int64(3), object(3)
memory usage: 20.6+ KB


In [91]:
# The Summary Statistics:

schools.describe()

average_math  average_reading  average_writing  percent_tested
count    375.000000       375.000000       375.000000      355.000000
mean     432.944000       424.504000       418.458667       64.976338
std       71.952373        61.881069        64.548599       18.747634
min      317.000000       302.000000       284.000000       18.500000
25%      386.000000       386.000000       382.000000       50.950000
50%      415.000000       413.000000       403.000000       64.800000
75%      458.500000       445.000000       437.500000       79.600000
max      754.000000       697.000000       693.000000      100.000000

- The average math mark for all the schools: 432.944
- The average reading mark for all the schools: 424.504
- The average writing mark for all the schools : 418.458


We will now now explore these marks per school.



In [92]:
# Let us analyse the math results for each school.

min_math_mark = 0.8*800


# Comments: If a school has an average mark greater than this, then its a best math performing school.

# Subsetting schools with the required mark
best_math_schools = schools[ schools['average_math'] >= min_math_mark]

# Sorting the schools average math mark.
best_math_schools = best_math_schools.sort_values(by = 'average_math', ascending = False)

#Subsetting the specified columns from our relevant dataframe
best_math_schools = best_math_schools[['school_name', 'average_math']]

best_math_schools




school_name  average_math
88                              Stuyvesant High School           754
170                       Bronx High School of Science           714
93                 Staten Island Technical High School           711
365  Queens High School for the Sciences at York Co...           701
68   High School for Mathematics, Science, and Engi...           683
280                     Brooklyn Technical High School           682
333                        Townsend Harris High School           680
174  High School of American Studies at Lehman College           669
0    New Explorations into Science, Technology and ...           657
45                       Eleanor Roosevelt High School           641

Some Analysis For the Above:



In [93]:
print("The best math performing school:", best_math_schools.iloc[0, 0])
print("There are {} best math performing schools".format(len(best_math_schools)))

The best math performing school: Stuyvesant High School
There are 10 best math performing schools


What are the top 10 performing schools based on the combined SAT scores?

In [98]:
# For this we will have to create a new column named : "total_SAT"

schools["total_SAT"] = (schools["average_math"] + schools["average_reading"] + schools["average_writing"])

schools.head()

school_name  ... total_SAT
0  New Explorations into Science, Technology and ...  ...      1859
1                               Essex Street Academy  ...      1193
2                       Lower Manhattan Arts Academy  ...      1261
3    High School for Dual Language and Asian Studies  ...      1529
4      Henry Street School for International Studies  ...      1197

[5 rows x 8 columns]

In [101]:
# The top ten schools based on the combined SAT scores.

schools_sorted = schools.sort_values('total_SAT', ascending = False)

top_10_schools = schools_sorted[['school_name', 'total_SAT']].iloc[0:10, ]

top_10_schools

school_name  total_SAT
88                              Stuyvesant High School       2144
170                       Bronx High School of Science       2041
93                 Staten Island Technical High School       2041
174  High School of American Studies at Lehman College       2013
333                        Townsend Harris High School       1981
365  Queens High School for the Sciences at York Co...       1947
5                       Bard High School Early College       1914
280                     Brooklyn Technical High School       1896
45                       Eleanor Roosevelt High School       1889
68   High School for Mathematics, Science, and Engi...       1889

Which single borough has the largest standard deviation in the combined SAT score?

In [102]:
borough_sorted = round(schools.groupby('borough')["total_SAT"].agg(['count', 'mean', 'std']).sort_values('std', ascending = False), 2)


borough_sorted= borough_sorted.rename(columns ={'count':'num_schools',
                                               'mean':'average_SAT',
                                               'std':'std_SAT'})

largest_std_dev = borough_sorted.iloc[[0], :]

largest_std_dev



num_schools  average_SAT  std_SAT
borough                                     
Manhattan           89      1340.13   230.29